test/Jwt/Constraint/HasValidExpirationTest.php

```diff
@@ -0,0 +1,139 @@
+<?php declare(strict_types=1);
+/**
+ * This file is part of pmg/assertion-grant.
+ *
+ * Copyright (c) PMG <https://www.pmg.com>
+ *
+ * For full copyright and license information, please view the LICENSE file
+ * that was distributed with this source code.
+ */
+
+namespace PMG\AssertionGrant\Test\Jwt\Constraint;
+
+use DateInterval;
+use DateTimeImmutable;
+use Lcobucci\Clock\FrozenClock;
+use Lcobucci\JWT\Configuration;
+use Lcobucci\JWT\Token;
+use Lcobucci\JWT\Validation\RequiredConstraintsViolated;
+use PMG\AssertionGrant\Jwt\Constraint\HasValidExpiration;
+use PMG\AssertionGrant\Test\Jwt\JwtTestCase;
+
+class HasValidExpirationTest extends JwtTestCase
+{
+    private Configuration $jwtConfig;
+    private DateTimeImmutable $now;
+    private FrozenClock $clock;
+    private DateInterval $maxTtl;
+
+    public function testNonPlainTokensCauseErrors() : void
+    {
+        $this->expectException(RequiredConstraintsViolated::class);
+        $this->expectExceptionMessage('only plaintext tokens are supported');
+
+        $token = $this->createMock(Token::class);
+
+        $this->jwtConfig->validator()->assert($token, new HasValidExpiration(
+            $this->clock,
+            $this->maxTtl
+        ));
+    }
+
+    public function testCurrentTimeIsUsedToValidatedExpirationIfNbfAndIssuedAtIsMissing() : void
+    {
+        $token = $this->jwtConfig->builder()
+            ->expiresAt($this->now->add(new DateInterval('PT30M')))
+            ->getToken($this->jwtConfig->signer(), $this->jwtConfig->signingKey());
+
+        $ok = $this->jwtConfig->validator()->validate($token, new HasValidExpiration(
+            $this->clock,
+            $this->maxTtl
+        ));
+
+        $this->assertTrue($ok);
+    }
+
+    public function testNotBeforeClaimIsUsedToCheckExpirationTimeIfPresent() : void
+    {
+        // 30 minute window between nbf and expiration
```

Comment:
what is `nbf`?

Reply:
the not before claim: https://www.iana.org/assignments/jwt/jwt.xhtml#claims

Reply:
Oh hi @ahuggins